### PR TITLE
Fix error with nested context manager.

### LINKED ISSUE #206: Issues with nested context manager when using cached object
Reproducer:

```python
class Backend:
    __ua_domain__ = "ua_bug"

    @staticmethod
    @ua.wrap_single_convertor
    def __ua_convert__(value, dispatch_type, coerce):
        return value

    @staticmethod
    def __ua_function__(func, args, kwargs):
        pass

# Should just create the context manager, not set the backend
be = ua.set_backend(Backend)


with be:
    with be:
        pass
```

Error message:

```
Traceback (most recent call last):
     ...
RuntimeError: Found invalid context state while in __exit__
```

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #206

### CODE REVIEW

**Comment by hameerabbasi** [comment]
It seems `test_invalid` is causing the segfault, even though it does pass. The Python process tries to `decref` something too much at one point, I think.

File: uarray/_uarray_dispatch.cxx [hunk 1]
```diff
@@ -309,14 +310,21 @@ class context_helper
...
-      if (!success)
+      bool success = true;
+      if (sizes_.size() == 0 || backends_->size() != sizes_.back()) {
         PyErr_SetString(PyExc_RuntimeError,
                         "Found invalid context state while in __exit__");
+        success = false;
+      }
+      try {
+        sizes_.pop_back();
+        backends_->pop_back();
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: `pop_back` doesn't allocate so no need for the `try`-`catch`.

File: uarray/_uarray_dispatch.cxx [hunk 2]
```diff
@@ -309,14 +310,21 @@ class context_helper
...
   bool exit()
     {
-      bool success = (enter_size_ + 1 == backends_->size()
-                      && backends_->back() == new_backend_);
-      if (enter_size_ < backends_->size())
-        backends_->resize(enter_size_);
-
-      if (!success)
+      bool success = true;
+      if (sizes_.size() == 0 || backends_->size() != sizes_.back()) {
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: If `sizes_.size() == 0`then calling `pop_back` is undefined behaviour.

Comment (hameerabbasi) [reply uarray/_uarray_dispatch.cxx]: What do you suggest we do in this case? I'm inclined to raise a `SystemError`.

Comment (peterbell10) [reply uarray/_uarray_dispatch.cxx]: What was wrong with the way it was done before?
```
      if (enter_size_ < backends_->size())
        backends_->resize(enter_size_);

```

Comment (hameerabbasi) [reply uarray/_uarray_dispatch.cxx]: `enter_size_` may not be defined if `enter_` was not called.

File: uarray/_uarray_dispatch.cxx [hunk 3]
```diff
@@ -309,14 +310,20 @@ class context_helper
...
-      bool success = (enter_size_ + 1 == backends_->size()
-                      && backends_->back() == new_backend_);
-      if (enter_size_ < backends_->size())
-        backends_->resize(enter_size_);
-
-      if (!success)
+      bool success = true;
+      if (sizes_.size() == 0) {
+        PyErr_SetString(PyExc_SystemExit,
+                        "Found invalid context state while in __exit__");
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
                        "__exit__ call has no matching __enter__");
```

File: uarray/_uarray_dispatch.cxx [hunk 4]
```diff
@@ -309,14 +310,20 @@ class context_helper
...
   bool exit()
     {
-      bool success = (enter_size_ + 1 == backends_->size()
-                      && backends_->back() == new_backend_);
-      if (enter_size_ < backends_->size())
-        backends_->resize(enter_size_);
-
-      if (!success)
+      bool success = true;
+      if (sizes_.size() == 0) {
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
      if (sizes_.empty()) {
```

File: uarray/_uarray_dispatch.cxx (hunk 2 above)
Comment (peterbell10) [reply uarray/_uarray_dispatch.cxx]: Ah of course, `sizes_.size()` was making me think of no backends, rather than no matching `__enter__` call.

**Pull request merged**